pkg/controller/jobframework/setup.go

```diff
@@ -71,25 +76,21 @@ func (m *integrationManager) setupControllers(mgr ctrl.Manager, log logr.Logger,
 			if err != nil {
 				return fmt.Errorf("%s: %w: %w", fwkNamePrefix, errFailedMappingResource, err)
 			}
-			if _, err = mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version); err != nil {
+			if _, err := getRESTMapping(mgr, gvk); err != nil {
 				if !meta.IsNoMatchError(err) {
 					return fmt.Errorf("%s: %w", fwkNamePrefix, err)
 				}
 				logger.Info("No matching API in the server for job framework, skipped setup of controller and webhook")
+				go waitForAPI(ctx, mgr, log, gvk, options.IntegrationRetryInterval.Duration, func() {
```

Comment:
I'm ok with this as a quick fix, but I'm wondering if it would be better / feasible to have a watch on the CRDs (an analog of `kubectl get crd -w --output-watch-events`), and install the Controller and webhook when we see the ADDED event for the CRD in question. Unless I'm missing something, this could work without any user-facing API.

Reply:
Previously, that was the original idea. I was going to use [`apiextensions`](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions) to watch CRDs directly, however I was suggested to use the restmapper. I suppose it is more straightforward and it's also how we've been checking for their availability. This is from [a draft PR](https://github.com/opendatahub-io/kueue/pull/33/commits/5bec43389f43dc7a71d644d4efa9d6f456f77fc4#diff-9b0c3819f5a2c007ac714ec151937ce08cc2b768527cc28d20580a57f47ca5a7R155) from some time ago (might need more work). A follow-up can be made on this.

Reply:
> I was going to use [apiextensions](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions) to watch CRDs directly, however I was suggested to use the restmapper.

Thank you for the context. I still think the current approach is suboptimal, whatever is the interval, because it means periodic polling, which takes goroutines and results in spammy logging. 

So, I think we would like to use a watch for CRDs  to guide the RestMapper check. So only once we discover the new CRD, then we start waiting for the rest mapper to be filled. Then, the waiting intervals could be much smaller (and don't need to be configurable).

One way to the watcher would be to have a controller-runtime reconciler for the CRD type, similar to the one for [ResourceFlavor](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/resourceflavor_controller.go). I imagine we would use `WithEventFilter` to only listen for the ADD events. Then trigger reconciler. The Reconcile function would actually call the rest mapper. If succeeds, then done, if fails, then it will retry with exponential backoff, but this is already handled by controller-runtime.

Reply:
Then we could also investigate if we can do the symmetric operation - when the CRD is removed (DELETE event), can we unregister the controller and the webhook.

Reply:
@mimowo having a separate controller watch for a CRD, and then send in the event through a channel to trigger the start of the main controller does follow the event driven principle and reduces the delay that comes by polling at the specified interval. But the reason for having it this way, was to introduce as minimum complexity as possible upstream, as this would be an occurrence only once before the start of the main controller. 

> One way to the watcher would be to have a controller-runtime reconciler for the CRD type, similar to the one for [ResourceFlavor](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/core/resourceflavor_controller.go).

Looking at the code (pl correct me if I'm wrong), each of the watchers specified in the ResourceFlavour, receive events through channels created in their respective reconcilers (AdmissionCheck, LocalClusterQueue, ClusterQueue) - https://github.com/kubernetes-sigs/kueue/blob/58ef31e6ef674db0aa04cb587d41862570ad1d12/pkg/controller/core/clusterqueue_controller.go#L89. But in this case, the list of CRDs is configurable, so having a reconciler for each type would not be scalable. One option here, would be to have a single reconciler, and have a dynamic restMapper (with a dynamic Client) to watch for all the expected CRD types and send in a signal to start the core controller. This is going to be very similar (without the creation of separate reconciler) to @ChristianZaccaria option (1) as mentioned in the issue.

>  when the CRD is removed (DELETE event), can we unregister the controller and the webhook.

This touches the part of dynamically starting/stopping the reconcilers, which unfortunately is not natively supported by controller-runtime yet (https://github.com/kubernetes-sigs/controller-runtime/pull/2159). There are ways to do workarounds on this - like cancelling the context, stopping the informers (which may not be trivial), registering a stopChn with `manager` to signal a shutdown. That being said - is this the behaviour we would expect? Since the option to provide integrations is provided by the user through config API at the start of the Kueue controller, and iiuc there is no way to signal that the integration is no longer required - I would expect the Kueue controller to fail and error if the CRD is removed, rather than us unregistering the respective controller?

Reply:
> But the reason for having it this way, was to introduce as minimum complexity as possible upstream, as this would be an occurrence only once before the start of the main controller.

I agree with reducing the complexity, but being uncertain about the long-term solution does not make me comfortable extending the API surface, because likely we will need to deprecate it one day. Having hard-coded 20s for the current solution is fine for me.

> But in this case, the list of CRDs is configurable, so having a reconciler for each type would not be scalable.

I was thinking we could have the just a single reconciler with a single watch for the `CustomResourceDefinition` kind. Then, based on the object, we could read the details of the new CRD installed: https://github.com/kubernetes/kubernetes/blob/ceb58a4dbc671b9d0a2de6d73a1616bc0c299863/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go#L36-L44, and lookup the RESTMapper.

>  I would expect the Kueue controller to fail and error if the CRD is removed, rather than us unregistering the respective controller?

I'm not sure, maybe it should be configurable. Anyway, I think we can park this particular decision for now. 

In summary, I would suggest to research a little more the reconciler based approach, if there are obstacles, we can go with the current one, but I wouldn't extend the API surface for it.

Reply:
> I agree with reducing the complexity, but being uncertain about the long-term solution does not make me comfortable extending the API surface, because likely we will need to deprecate it one day. Having hard-coded 20s for the current solution is fine for me.

Agreed, seems fair enough to not extend an API surface especially given that it's not going to impact the Kueue functionality largely and is used only during setup. 

> I was thinking we could have the just a single reconciler with a single watch for the CustomResourceDefinition kind. 

I'm a bit skeptical with controller approach because - Controllers are useful when there is a need to manage the state of the resource through continuous reconciliation, in addition to being able to integrate with other controllers (which is the use-case here). The design mentioned above complies perfectly with the current Watcher approach, wherein say the `ClusterQueueReconciler` is managing the state of `ClusterQueue` resource apart from being able to send a signal to another controller. In this case, we are not intending to manage the lifecycle of any of the CRDs (there are respective operators for each one of them), nor are going to manage the CRs created for them throughout the existence of Kueue's core controller. 

Having a separate controller, apart from its benefits - is also going to give us the overhead of caching these CRDs, spinning up informers for them (which currently cannot be stopped as of c-r design) consuming memory and CPU resources, while we are going to use it only during setup. Imo having a controller for this seems to be an overkill. Given we are not going to deal with managing the state of any of the resources - a simple Watch makes more sense, given its going to be lightweight,  where we can just close the channel once we have received the event we want and have started the core controller and also follows the event driven approach rather than polling.

Reply:
@mimowo would you be comfortable with making the `retryInterval` constant in the current design, and following up further on the watch based approach as suggested by @ChristianZaccaria in [here](https://github.com/opendatahub-io/kueue/pull/33/commits/5bec43389f43dc7a71d644d4efa9d6f456f77fc4#diff-9b0c3819f5a2c007ac714ec151937ce08cc2b768527cc28d20580a57f47ca5a7R155).

Reply:
Yes, I would also like you folks to consider adding an integration test for it on the way.

Reply:
One more thing worth considering also is an exponential backoff instead of fixed intervals for for calling RESTmapper (say from 1s to 1min). I believe generally the RESTMapper would be updated very fast, but maybe for an overloaded system it makes sense to throttle the attempts. WDYT?

Reply:
> I believe generally the RESTMapper would be updated very fast, but maybe for an overloaded system it makes sense to throttle the attempts.

Yeah, polling it quickly might indeed overload the apiserver. My only concern here would be the responsiveness of the overall system with us trying exponential backoff. If we can have a reasonable limit on the maximum backoff time, then it should be good.

Reply:
Reconcilers use exponential rate limiters: https://github.com/kubernetes/client-go/blob/fe54892e0cc44385de5b5300985bd27218dd2c23/util/workqueue/default_rate_limiters.go#L112C6-L115, so from 1ms to 1000s by default. We can try that and see how it works in practice, I'm ok for adjust the lower end to say 100ms (or 1s) if it turns out there are too many failed attempts.

Reply:
Exponential rate limiters sound like a great idea, will be implemented in this PR. Thanks! We could re-discuss to agree on a lower and upper end values.